lib/PuppeteerSharp/CDPSession.cs

```diff
@@ -184,22 +184,32 @@ internal void OnMessage(ConnectionResponse obj)
         {
             var id = obj.Id;
 
-            if (id.HasValue && _callbacks.TryRemove(id.Value, out var callback))
+            if (id.HasValue)
             {
-                if (obj.Error != null)
+                if (_callbacks.TryRemove(id.Value, out var callback))
                 {
-                    callback.TaskWrapper.TrySetException(new MessageException(callback, obj.Error));
-                }
-                else
-                {
-                    callback.TaskWrapper.TrySetResult(obj.Result);
+                    // This is a response to a SendAsync. Handle the callback async, since (a) we can, and (b)
+                    // to avoid a situation where the continuation tries to call SendAsync again thus creating
+                    // a deadlock.
+                    Task.Run(() =>
+                    {
+                        if (obj.Error != null)
+                        {
+                            callback.TaskWrapper.TrySetException(new MessageException(callback, obj.Error));
+                        }
+                        else
+                        {
+                            callback.TaskWrapper.TrySetResult(obj.Result);
+                        }
+                    });
                 }
             }
             else
             {
                 var method = obj.Method;
                 var param = obj.Params?.ToObject<ConnectionResponseParams>();
 
+                // Always call sync, since we need message (event) order to be determinate.
                 MessageReceived?.Invoke(this, new MessageEventArgs
```

Comment:
Note that I left this unchanged since otherwise Chrome events would be raised out of order.